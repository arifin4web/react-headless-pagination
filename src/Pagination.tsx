import React, { FC } from "react";
import classNames from "classnames";
import usePagination from "./hooks/usePagination";
import {
  IPagination,
  IPaginationProps,
  ButtonProps,
  PageButtonProps,
} from "./Pagination.d";

export const PrevButton: FC<ButtonProps> = ({
  className,
  children,
  ...buttonProps
}) => {
  const pagination: IPagination = React.useContext(PaginationContext);
  const previous = () => {
    if (pagination.currentPage + 1 > 1) {
      pagination.setCurrentPage(pagination.currentPage - 1);
    }
  };

  return (
    <button
      className={className}
      {...buttonProps}
      onClick={() => previous()}
      disabled={pagination.currentPage === 0}
    >
      {children}
    </button>
  );
};

export const NextButton: FC<ButtonProps> = ({
  className,
  children,
  ...buttonProps
}) => {
  const pagination: IPagination = React.useContext(PaginationContext);
  const next = () => {
    if (pagination.currentPage + 1 < pagination.pages.length) {
      pagination.setCurrentPage(pagination.currentPage + 1);
    }
  };

  return (
    <button
      className={className}
      {...buttonProps}
      onClick={() => next()}
      disabled={pagination.currentPage === pagination.pages.length - 1}
    >
      {children}
    </button>
  );
};

interface ITruncableElementProps {
  prev?: boolean;
}

export const TruncableElement: FC<ITruncableElementProps> = ({ prev }) => {
  const pagination: IPagination = React.useContext(PaginationContext);

  const {
    isPreviousTruncable,
    isNextTruncable,
    truncableText = "...",
    truncableClassName = "",
  } = pagination;

  return (isPreviousTruncable && prev === true) ||
    (isNextTruncable && !prev) ? (
    <span className={truncableClassName}>{truncableText}</span>
  ) : null;
};

export const PageButton: FC<PageButtonProps> = ({
  className,
  activeClassName = "",
  inactiveClassName = "",
}) => {
  const pagination: IPagination = React.useContext(PaginationContext);

  const renderPageButton = (page: number) => (
    <span
      key={page}
      className={classNames(className, {
        [activeClassName]: pagination.currentPage + 1 === page,
        [inactiveClassName]: pagination.currentPage + 1 !== page,
      })}
      onClick={() => pagination.setCurrentPage(page - 1)}
    >
      {page}
    </span>
  );

  return (
    <>
      {pagination.previousPages.map(renderPageButton)}
      <TruncableElement prev />
      {pagination.middlePages.map(renderPageButton)}
      <TruncableElement />
      {pagination.nextPages.map(renderPageButton)}
    </>
  );
};

const defaultState: IPagination = {
  currentPage: 0,
  setCurrentPage: () => {},
  truncableText: "...",
  truncableClassName: "",
  pages: [],
  hasPreviousPage: false,
  hasNextPage: false,
  previousPages: [],
  isPreviousTruncable: false,
  middlePages: [],
  isNextTruncable: false,
  nextPages: [],
};

const PaginationContext: React.Context<IPagination> =
  React.createContext(defaultState);

export const Pagination = (paginationProps: IPaginationProps) => {
  const pagination = usePagination(paginationProps);

  return (
    <PaginationContext.Provider value={pagination}>
      <div className={paginationProps.className}>
        {paginationProps.children}
      </div>
    </PaginationContext.Provider>
  );
};

Pagination.PrevButton = PrevButton;
Pagination.NextButton = NextButton;
Pagination.PageButton = PageButton;
