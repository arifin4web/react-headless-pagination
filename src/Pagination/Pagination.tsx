import React from "react";
import classNames from "classnames";
import usePagination from "../Hooks/usePagination";
import {
  IPagination,
  IPaginationProps,
  ButtonProps,
  PageButtonProps,
} from "../types/types";

export const PrevButton = ({
  className,
  children,
  dataTestId,
  as = <button />,
  ...buttonProps
}: ButtonProps) => {
  const pagination = React.useContext(PaginationContext);
  const previous = () => {
    if (pagination.currentPage + 1 > 1) {
      pagination.setCurrentPage(pagination.currentPage - 1);
    }
  };

  return (
    <as.type
      {...buttonProps}
      {...as.props}
      className={classNames(className, as.props.className)}
      onClick={() => previous()}
      disabled={pagination.currentPage === 0}
      data-testid={dataTestId}
    >
      {children}
    </as.type>
  );
};

export const NextButton = ({
  className,
  children,
  dataTestId,
  as = <button />,
  ...buttonProps
}: ButtonProps) => {
  const pagination = React.useContext(PaginationContext);
  const next = () => {
    if (pagination.currentPage + 1 < pagination.pages.length) {
      pagination.setCurrentPage(pagination.currentPage + 1);
    }
  };

  return (
    <as.type
      {...buttonProps}
      {...as.props}
      className={classNames(className, as.props.className)}
      onClick={() => next()}
      disabled={pagination.currentPage === pagination.pages.length - 1}
      data-testid={dataTestId}
    >
      {children}
    </as.type>
  );
};

type ITruncableElementProps = {
  prev?: boolean;
};

const TruncableElement = ({ prev }: ITruncableElementProps) => {
  const pagination: IPagination = React.useContext(PaginationContext);

  const {
    isPreviousTruncable,
    isNextTruncable,
    truncableText,
    truncableClassName,
  } = pagination;

  return (isPreviousTruncable && prev === true) ||
    (isNextTruncable && !prev) ? (
    <li className={truncableClassName || undefined}>{truncableText}</li>
  ) : null;
};

export const PageButton = ({
  as = <a />,
  className,
  dataTestIdActive,
  dataTestIdInactive,
  activeClassName,
  inactiveClassName,
}: PageButtonProps) => {
  const pagination: IPagination = React.useContext(PaginationContext);

  const renderPageButton = (page: number) => (
    <li key={page}>
      <as.type
        data-testid={
          classNames({
            [`${dataTestIdActive}`]:
              dataTestIdActive && pagination.currentPage + 1 === page,
            [`${dataTestIdInactive}-${page}`]:
              dataTestIdActive && pagination.currentPage + 1 !== page,
          }) || undefined
        }
        tabIndex={0}
        onKeyPress={(event: React.KeyboardEvent) => {
          if (event.key === "Enter") {
            pagination.setCurrentPage(page - 1);
          }
        }}
        onClick={() => pagination.setCurrentPage(page - 1)}
        className={classNames(
          className,
          pagination.currentPage + 1 === page
            ? activeClassName
            : inactiveClassName,
        )}
        {...as.props}
      >
        {page}
      </as.type>
    </li>
  );

  return (
    <>
      {pagination.previousPages.map(renderPageButton)}
      <TruncableElement prev />
      {pagination.middlePages.map(renderPageButton)}
      <TruncableElement />
      {pagination.nextPages.map(renderPageButton)}
    </>
  );
};

const defaultState: IPagination = {
  currentPage: 0,
  setCurrentPage: () => {},
  truncableText: "...",
  truncableClassName: "",
  pages: [],
  hasPreviousPage: false,
  hasNextPage: false,
  previousPages: [],
  isPreviousTruncable: false,
  middlePages: [],
  isNextTruncable: false,
  nextPages: [],
};

const PaginationContext: React.Context<IPagination> =
  React.createContext<IPagination>(defaultState);

export const Pagination = ({
  dataTestId,
  ...paginationProps
}: IPaginationProps & { dataTestId?: string }) => {
  const pagination = usePagination(paginationProps);

  return (
    <PaginationContext.Provider value={pagination}>
      <div className={paginationProps.className} data-testid={dataTestId}>
        {paginationProps.children}
      </div>
    </PaginationContext.Provider>
  );
};

Pagination.PrevButton = PrevButton;
Pagination.NextButton = NextButton;
Pagination.PageButton = PageButton;
