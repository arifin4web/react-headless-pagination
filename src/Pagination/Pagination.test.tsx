import React from "react";
import { FiArrowLeft, FiArrowRight } from "react-icons/fi";
import { render } from "@testing-library/react";
import { Pagination } from "./Pagination";
import { IPaginationProps } from "./Pagination.d";

const setupPagination = ({
  currentPage = 1,
  setCurrentPage = () => {},
  className = "",
  truncableText = "...",
  truncableClassName = "",
  totalPages = 10,
  edgePageCount = 1,
  middlePagesSiblingCount = 2,
  children = <></>,
}: Partial<IPaginationProps>) =>
  render(
    <Pagination
      currentPage={currentPage}
      setCurrentPage={setCurrentPage}
      className={className}
      truncableText={truncableText}
      truncableClassName={truncableClassName}
      totalPages={totalPages}
      edgePageCount={edgePageCount}
      middlePagesSiblingCount={middlePagesSiblingCount}
    >
      {children}
    </Pagination>,
  );

describe("Pagination", () => {
  it("renders correctly with basic setup of prevButton, pageButton and nextButton", () => {
    const { asFragment } = setupPagination({
      children: (
        <>
          <Pagination.PrevButton className="">Previous</Pagination.PrevButton>

          <div className="flex items-center justify-center flex-grow">
            <Pagination.PageButton
              activeClassName=""
              inactiveClassName=""
              className=""
            />
          </div>

          <Pagination.NextButton className="">Next</Pagination.NextButton>
        </>
      ),
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it("renders correctly with advanced setup of prevButton, pageButton and nextButton using Tailwind", () => {
    const { asFragment } = setupPagination({
      className: "flex items-center w-full h-10 text-sm select-none",
      truncableClassName: "w-10 px-0.5 text-center",
      children: (
        <>
          <Pagination.PrevButton className="flex items-center mr-2 text-gray-500 cursor-pointer hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none">
            <FiArrowLeft size={20} className="mr-3" />
            Previous
          </Pagination.PrevButton>

          <div className="flex items-center justify-center flex-grow">
            <Pagination.PageButton
              activeClassName="bg-primary-50 dark:bg-opacity-0 text-primary-600 dark:text-white"
              inactiveClassName="text-gray-500"
              className={
                "flex items-center justify-center h-10 w-10 rounded-full cursor-pointer"
              }
            />
          </div>

          <Pagination.NextButton className="flex items-center mr-2 text-gray-500 opacity-50 hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none">
            Next
            <FiArrowRight size={20} className="ml-3" />
          </Pagination.NextButton>
        </>
      ),
    });

    expect(asFragment()).toMatchSnapshot();
  });
});
