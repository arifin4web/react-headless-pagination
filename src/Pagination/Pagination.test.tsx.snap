// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly with advanced setup of prevButton, pageButton and nextButton using Tailwind 1`] = `
<DocumentFragment>
  <div
    class="flex items-center w-full h-10 text-sm select-none"
  >
    <button
      class="flex items-center mr-2 text-gray-500 cursor-pointer hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none"
    >
      <svg
        class="mr-3"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="19"
          x2="5"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 19 5 12 12 5"
        />
      </svg>
      Previous
    </button>
    <div
      class="flex items-center justify-center flex-grow"
    >
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
      >
        1
      </span>
      <span
        class="w-10 px-0.5 text-center"
      >
        ...
      </span>
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
      >
        4
      </span>
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
      >
        5
      </span>
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer bg-primary-50 dark:bg-opacity-0 text-primary-600 dark:text-white"
      >
        6
      </span>
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
      >
        7
      </span>
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
      >
        8
      </span>
      <span
        class="w-10 px-0.5 text-center"
      >
        ...
      </span>
      <span
        class="flex items-center justify-center h-10 w-10 rounded-full cursor-pointer text-gray-500"
      >
        10
      </span>
    </div>
    <button
      class="flex items-center mr-2 text-gray-500 opacity-50 hover:text-gray-600 dark:hover:text-gray-200 focus:outline-none"
    >
      Next
      <svg
        class="ml-3"
        fill="none"
        height="20"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <line
          x1="5"
          x2="19"
          y1="12"
          y2="12"
        />
        <polyline
          points="12 5 19 12 12 19"
        />
      </svg>
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination renders correctly with basic setup of prevButton, pageButton and nextButton 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      data-testid="prev"
    >
      Previous
    </button>
    <div
      class="flex items-center justify-center flex-grow"
    >
      <span
        data-testid="inactive-page-button-1"
      >
        1
      </span>
      <span>
        ...
      </span>
      <span
        data-testid="inactive-page-button-4"
      >
        4
      </span>
      <span
        data-testid="inactive-page-button-5"
      >
        5
      </span>
      <span
        data-testid="active-page-button"
      >
        6
      </span>
      <span
        data-testid="inactive-page-button-7"
      >
        7
      </span>
      <span
        data-testid="inactive-page-button-8"
      >
        8
      </span>
      <span>
        ...
      </span>
      <span
        data-testid="inactive-page-button-10"
      >
        10
      </span>
    </div>
    <button
      data-testid="next"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;

exports[`Pagination renders correctly with different truncable text 1`] = `
<DocumentFragment>
  <div
    class=""
  >
    <button
      data-testid="prev"
    >
      Previous
    </button>
    <div
      class="flex items-center justify-center flex-grow"
    >
      <span
        data-testid="inactive-page-button-1"
      >
        1
      </span>
      <span>
        ---
      </span>
      <span
        data-testid="inactive-page-button-4"
      >
        4
      </span>
      <span
        data-testid="inactive-page-button-5"
      >
        5
      </span>
      <span
        data-testid="active-page-button"
      >
        6
      </span>
      <span
        data-testid="inactive-page-button-7"
      >
        7
      </span>
      <span
        data-testid="inactive-page-button-8"
      >
        8
      </span>
      <span>
        ---
      </span>
      <span
        data-testid="inactive-page-button-10"
      >
        10
      </span>
    </div>
    <button
      data-testid="next"
    >
      Next
    </button>
  </div>
</DocumentFragment>
`;
